proxy/core/src/test/scala/io/cloudstate/proxy/crdt/LWWRegisterCrdtEntitySpec.scala

```diff
@@ -97,12 +98,16 @@ class LWWRegisterCrdtEntitySpec extends AbstractCrdtEntitySpec {
       expectNoMessage(200.millis)
 
       val cid2 = sendAndExpectCommand("cmd", command)
+
+      // Walsh: failing line
       sendAndExpectReply(cid2, updateRegister(element3, clock = CrdtClock.CUSTOM, customClockValue = start))
```

Comment:
What is the failing message?

Reply:
[info] - should handle change deltas with custom clock *** FAILED ***
[info]   java.lang.AssertionError: assertion failed: timeout (3 seconds) during expectMsgClass waiting for class io.cloudstate.proxy.entity.UserFunctionReply
[info]   at scala.Predef$.assert(Predef.scala:223)
[info]   at akka.testkit.TestKitBase.expectMsgClass_internal(TestKit.scala:529)
[info]   at akka.testkit.TestKitBase.expectMsgType(TestKit.scala:501)
[info]   at akka.testkit.TestKitBase.expectMsgType$(TestKit.scala:500)
[info]   at akka.testkit.TestKit.expectMsgType(TestKit.scala:929)
[info]   at io.cloudstate.proxy.crdt.AbstractCrdtEntitySpec.doSendAndExpectReply(AbstractCrdtEntitySpec.scala:180)
[info]   at io.cloudstate.proxy.crdt.AbstractCrdtEntitySpec.sendAndExpectReply(AbstractCrdtEntitySpec.scala:134)
[info]   at io.cloudstate.proxy.crdt.LWWRegisterCrdtEntitySpec.$anonfun$new$16(LWWRegisterCrdtEntitySpec.scala:103)
[info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[info]   ...
[info] - should handle change deltas with custom auto incrementing clock *** FAILED ***
[info]   java.lang.AssertionError: assertion failed: expected class akka.cluster.ddata.Replicator$UpdateSuccess, found class io.cloudstate.proxy.entity.UserFunctionReply (UserFunctionReply(Some(ClientAction(Failure(Failure(0,Failed to update CRDT at requested write consistency)))),List()))
[info]   at scala.Predef$.assert(Predef.scala:223)
[info]   at akka.testkit.TestKitBase.expectMsgClass_internal(TestKit.scala:530)
[info]   at akka.testkit.TestKitBase.expectMsgType(TestKit.scala:501)
[info]   at akka.testkit.TestKitBase.expectMsgType$(TestKit.scala:500)
[info]   at akka.testkit.TestKit.expectMsgType(TestKit.scala:929)
[info]   at io.cloudstate.proxy.crdt.AbstractCrdtEntitySpec.update(AbstractCrdtEntitySpec.scala:101)
[info]   at io.cloudstate.proxy.crdt.LWWRegisterCrdtEntitySpec.$anonfun$new$18(LWWRegisterCrdtEntitySpec.scala:113)
[info]   at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)

Reply:
@jroper Do you have any idea why this could happen?

Reply:
Without looking too carefully at it, I'd guess something has changed slightly in the CRDT protocol - or maybe even hasn't changed, but got faster leading to a race condition that's always existed being triggered. The assertion is expecting a reply from the CRDT actor, but what it's actually got is a message from the Akka distributed data replicator. That wasn't sent by the CRDT actor. The message from the CRDT actor probably was sent successfully, but it's likely the next message after this replicator message.

Something I've found with Akka testkit is that it's very easy to accidentally have incorrect ordering expectations, where you send a message to A, not expecting to hear anything from it, and then send a message to B, expecting it to reply with X, then you assert that the next message you receive is X, but in fact A decided that it wanted to talk to you and sent you Y, and so your assertion fails, even though B did does send you X, it's just that Y arrived first. There's no way in Akka test kit to say "expect the next message from B that I receive to be X", you can only say "expect X from any actor". The way to solve it is to use multiple test actors, one for talking to A, and one for talking to B. So then you have testActorA send to A, it doesn't matter what if anything that A sends back, you then have testActorB send to B, and you can assert that the reply is X, knowing the communication between testActorA and A won't interfere.

Reply:
@jroper There is no such thing as `from` (unless you discount sender, which is removed in 2.6?) What you can do, however, is `fishForMsg`.

Reply:
The issue is more of a UX issue, not an issue of what methods are available. The problem is, the UX is tuned to make it appear that you're interacting directly with actors. But you're not, you are interacting through another actor (a test probe). When you understand that you're interacting through a test probe, it makes sense that if you use the same test probe to talk to two different actors, the results of those conversations are going to interfere, instead, you need two different test probes. But the UX tries to hide the test probe from you, is if your test code is the test probe, and that's where the mistakes happen.

Reply:
I'll look at removing ImplicitSender in favor of TestProbes.

Reply:
I think I've got it narrowed down enough that it would be good to pair up with James on Zoom and use it as a learning opportunity. I don't see any likely causes in the test, except the use of Source.actorRef, which is deprecated.

// Walsh: this is the last thing that happens before the following is encountered:
  // [akka://test/user/entity1] Message [akka.actor.Status$Failure] to Actor[akka://test/user/entity1#2000201393]
  protected def createAndExpectInit(): Option[S] = {
    toUserFunction = TestProbe()
    entityDiscovery = TestProbe()
    entity = system.actorOf(
      CrdtEntity.props(
        new Crdt() {
          override def handle(in: Source[CrdtStreamIn, NotUsed]): Source[CrdtStreamOut, NotUsed] = {
            in.runWith(Sink.actorRef(toUserFunction.testActor, StreamClosed))
            Source.actorRef(100, OverflowStrategy.fail).mapMaterializedValue { actor =>
              fromUserFunction = actor
              NotUsed
            }
          }
        },
        CrdtEntity.Configuration(ServiceName, UserFunctionName, Timeout(10.minutes), 100, 10.minutes, 10.minutes),
        new EntityDiscovery {
          override def discover(in: ProxyInfo): Future[EntitySpec] = {
            entityDiscovery.testActor ! in
            Future.failed(new Exception("Not expecting discover"))
          }
          override def reportError(in: UserFunctionError): Future[Empty] = {
            entityDiscovery.testActor ! in
            Future.successful(Empty())
          }
        }
      ),
      entityId
    )

Reply:
@jroper I've looked at everything I can think of, I've removed the unnecessary implicit materializers wherever they can possibly impact the test I'm running and there is only one use of ImplicitSender that seems ok to me. This is where I'm failing now, as part of the LWWRegisterCrdtEntitySpec, the following code is in AbstractCrdtEntitySpec:
protected def doSendAndExpectReply(commandId: Long,
                                     action: CrdtStateAction.Action,
                                     writeConsistency: CrdtWriteConsistency): UserFunctionReply = {
    sendReply(commandId, action, writeConsistency)
    println(s"---------------------------------${receiveN(1, 10.seconds)}")
    expectMsgType[UserFunctionReply]
  }

No reply ever happens. James, any chance we can pair up on this so you can take a look and hopefully explain some things?